### DISCUSSION TITLE
"Signature has no signature creation time subpacket in its hashed area."

### DISCUSSION THREAD

**Discussion by bjansen** [pgpainless/pgpainless#357] [Q&A] [answered]
Hi,

I'm trying to decrypt a file that was encrypted by SecureZIP and I'm getting the following error:

```
org.pgpainless.exception.SignatureValidationException: Signature is not valid: Malformed signature. Signature has no signature creation time subpacket in its hashed area.
```

I'm aware that RFC4880 requires the presence of the signature creation time _as a hashed subpacket_. It seems a little odd to me that a commercial product like SecureZIP would make such a mistake. 

`gpg` doesn't really seem to care that there is no such subpacket, it decrypts the file just fine and states that the signature is correct (event with the `--rfc4880` flag!).

I'm trying to determine what my options are. I think I could check if the `OpenPgpMetadata` contains an (invalid) signature made with the expected key, and simply ignore this specific kind of error.

On the other hand, getting SecureZIP to fix their bug (if it's really a bug on their side) seems like a really long shot.

What do you PGP experts think?

**Comment by bjansen** [comment]
Oh, I have no clue if it's relevant to my problem, but if I put a breakpoint in `org.pgpainless.signature.subpackets.SignatureSubpacketsUtil#getSignatureCreationTime`, I can see that the method is called multiple times. Only the call made by `org.pgpainless.signature.consumer.SignatureVerifier#verifyOnePassSignature` seems to fail, the ones made before succeed.

**Comment by vanitasvitae** [comment]
> Oh, I have no clue if it's relevant to my problem, but if I put a breakpoint in org.pgpainless.signature.subpackets.SignatureSubpacketsUtil#getSignatureCreationTime, I can see that the method is called multiple times.

This is likely because the method is reused by signature validation of signatures on the decryption / verification key.

> gpg doesn't really seem to care that there is no such subpacket, it decrypts the file just fine and states that the signature is correct (event with the --rfc4880 flag!).

That's odd. What version of gpg are you using? The [interoperability test suite](https://tests.sequoia-pgp.org/#Detached_signature_with_Subpackets) indicates that gpg `2.4.0`, `2.2.40` and `1.4.23` would also reject signatures without creation time packet (although the tests are dealing with detached signatures, though the logic is likely the same?).

Currently there is no way of changing PGPainless' behavior to allow verification of such malformed signatures. Have you reported this issue to the SecureZIP developers?

**Comment by bjansen** [comment]
I've tried gpg 2.0.22 on Linux and gpg 2.4.0 on macOS.

Here's the relevant section of `gpg --list-packets`:

```
# off=552 ctb=c4 tag=4 hlen=2 plen=13 new-ctb
:onepass_sig packet: keyid 60FD8B1D57D1313A
	version 3, sigclass 0x00, digest 2, pubkey 1, last=1
# off=567 ctb=cb tag=11 hlen=3 plen=996 new-ctb
:literal data packet:
	mode t (74), created 1675328130, name="ZQUA0W/XP/TIKWK523/G0001V00",
	raw data: 963 bytes
# off=1566 ctb=89 tag=2 hlen=3 plen=533
:signature packet: algo 1, keyid 60FD8B1D57D1313A
	version 3, created 1675328130, md5len 5, sigclass 0x00
	digest algo 2, begin of digest ab ad
	data: [4096 bits]
```

Usually I see lines like `hashed subpkt … (sig created 2021-03-07)` but not on this file.

**Reply by vanitasvitae** [reply]
So it apparently does have a creation time subpacket (`created 1675328130`), but possibly it is not in the hashed area? Still, the test suite reports that gpg rejects those in detached signatures, so I wonder why it behaves differently for inline signatures...

**Comment by bjansen** [comment, marked as answer]
Wait I think I got it. pgpainless throws an error on a signature packet in [version **3**](https://www.rfc-editor.org/rfc/rfc4880#section-5.2.2), but apparently hashed creation times are only required in [version **4**](https://www.rfc-editor.org/rfc/rfc4880#section-5.2.3.4).

**Reply by vanitasvitae** [reply]
Ah, in that case I guess its a bug in PGPainless. Will check if I can fix that.

Thanks for reporting!

**Reply by bjansen** [reply]
I can try to submit a PR if you'd like. I saw a class named `PGPV3SignatureGenerator` in BC which should help set up a repro case.

**Reply by bjansen** [reply]
Here's my attempt: https://github.com/pgpainless/pgpainless/pull/358

**Reply by vanitasvitae** [reply]
I just merged your PR into main. The fix will be part of 1.4.4 once I get around to make another release. Perhaps I will backport it to 1.3.X as well.
Thanks!

**Reply by bjansen** [reply]
Thanks! Do you have an ETA for 1.4.4 (or 1.3.x)? This issue is quite annoying to us, but as a workaround I guess I can still analyze the `OpenPgpMetadata` to determine if a signature was rejected because of this bug.

**Reply by vanitasvitae** [reply]
I can draft a release tomorrow :)
